### PR TITLE
Fix some framework coding issues

### PR DESCRIPTION
**Pull request by ghost**
### Description (*)
Used static code analysis to find a range of issues in the Magento Framework code. Mostly missing or incorrect typehints, undefined properties and too many parameters for method calls.

### Fixed Issues (if relevant)

1. N/A

### Manual testing scenarios (*)

1. As it's all on the framework level not easy to manually test. Easiest way might be to take a look at the code in an IDE and confirm the typehints now resolve to classes where before they didn't.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: lib/internal/Magento/Framework/App/Config.php [hunk 1]
```diff
@@ -50,7 +50,7 @@ public function __construct(
      *
      * @param string $path
      * @param string $scope
-     * @param null|string $scopeCode
+     * @param null|int|string $scopeCode
```
Comment (fooman) [inline lib/internal/Magento/Framework/App/Config.php:53]:
```
            if (is_numeric($scopeCode) || $scopeCode === null) {
                $scopeCode = $this->scopeCodeResolver->resolve($scope, $scopeCode);
            }
```

File: lib/internal/Magento/Framework/Cache/Frontend/Decorator/Logger.php [hunk 2]
```diff
@@ -44,7 +44,7 @@ public function remove($identifier)
      */
     public function clean($mode = \Zend_Cache::CLEANING_MODE_ALL, array $tags = [])
     {
-        $result = parent::clean($mode, $tags, $mode);
+        $result = parent::clean($mode, $tags);
```
Comment (fooman) [inline lib/internal/Magento/Framework/Cache/Frontend/Decorator/Logger.php:47]: public function clean($mode = \Zend_Cache::CLEANING_MODE_ALL, array $tags = [])
    {
        return $this->_getFrontend()->clean($mode, $tags);
    }

**Comment by fooman** [comment]
Okay so looks like there are a range of other things to address that got triggered by my changes...

I am okay addressing those but some of what is to be fixed is not clear at all. For example what is meant by the below?

```
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------
 25 | ERROR | [x] Missing short description
 26 | ERROR | [x] There must be exactly one blank line before tags
----------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```
phpcbf does not fix anything for me. 

I also tried running  `vendor/bin/phpcs --standard=Magento2` with `magento/magento-coding-standard` against the changed file and that brings up nothing.

**Comment by orlangur** [comment]
@fooman,

> phpcbf does not fix anything for me

Those are buggy sniffs which I already reported. They used `addFixableError` by mistake instead of `addError`.

File: dev/tests/integration/framework/Magento/TestFramework/App/MutableScopeConfig.php [hunk 3]
```diff
@@ -25,7 +25,7 @@ class MutableScopeConfig implements MutableScopeConfigInterface
     /**
      * @param string $path
      * @param string $scopeType
-     * @param null $scopeCode
+     * @param null|int||string $scopeCode
```
Comment (orlangur) [inline dev/tests/integration/framework/Magento/TestFramework/App/MutableScopeConfig.php]: Why `||`?

**Review by orlangur — CHANGES_REQUESTED** [review]
Thanks @fooman, please provide steps for catching those errors.

Also, class cannot be removed even if it's not used, just deprecate it.

File: dev/tests/integration/framework/Magento/TestFramework/App/MutableScopeConfig.php (hunk 3 above)
Comment (fooman) [reply dev/tests/integration/framework/Magento/TestFramework/App/MutableScopeConfig.php]: thanks - fixed

**Comment by fooman** [comment]
Sorry had some issues with git but should all be sorted now. The latest commit fixes the code style issues that were triggered by my touching the files.

The only exception so far is 
```
FILE: ...ar/www/html/lib/internal/Magento/Framework/App/Utility/Files.php
----------------------------------------------------------------------
FOUND 0 ERRORS AND 2 WARNINGS AFFECTING 2 LINES
----------------------------------------------------------------------
 618 | WARNING | The use of function md5() is discouraged; use
     |         | improved hash functions (SHA-256, SHA-512 etc.)()
     |         | instead
 771 | WARNING | The use of function md5() is discouraged; use
     |         | improved hash functions (SHA-256, SHA-512 etc.)()
     |         | instead
----------------------------------------------------------------------
```

where I don't blindly want to add an ignore without someone else making a call on this one. The code change to `hash('sha256'` would be simple enough but I am not sure if this would be too disruptive on existing caches, upgrades and what not. On the other side if some one could confirm that the md5 here is not security relevant that be great.

**Comment by fooman** [comment]
> please provide steps for catching those errors.

I have started using phpstan in my personal extensions which very quickly brings up Magento core related issues. The output is quite noisy so have manually been going through small chunks to fix what looks odd. Command would be something like `phpstan analyse  path/to/code -l1`

**Comment by orlangur** [comment]
@fooman,

> The output is quite noisy so have manually been going through small chunks to fix what looks odd. Command would be something like `phpstan analyse path/to/code -l1`

Perfect, such info is enough. Are a lot of false-positives or what kind of noise?

> On the other side if some one could confirm that the `md5` here is not security relevant that be great.

Generally, I don't think this rule should be applied anywhere outside the cryptographic layer. Will check for a particular use case here.

Thanks, will have a look on latest changes tonight.

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-5407](https://jira.corp.magento.com/browse/ENGCOM-5407) has been created to process this Pull Request
:eight_spoked_asterisk: @sivaschenko, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Pull request merged**